src/SubHask/Algebra/Matrix.hs

```diff
@@ -0,0 +1,340 @@
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
+{-# OPTIONS_GHC -fno-warn-unused-binds #-}
+
+module SubHask.Algebra.Matrix
+    ( Matrix (..)
+    , unsafeToModuleM
+    , ValidMatrix
+    , MatrixField
+    , mmult
+    , transposeM
+    , row
+    , col
+    , (!!)
+    )
+    where
+
+import qualified Prelude as P
+
+import Data.Primitive hiding (sizeOf)
+import GHC.TypeLits
+import Control.Monad.Primitive
+import Control.Monad
+import qualified Data.Primitive as Prim
+import qualified Data.Foldable
+
+import qualified SubHask.Algebra.Vector as V
+import SubHask.Algebra
+import SubHask.Category
+import SubHask.Internal.Prelude
+
+import qualified Numeric.LinearAlgebra
+
+{-# INLINE extendDimensions #-}
+extendDimensions :: Int -> Int
+extendDimensions = roundUpToNearest 4 -- i+4-i`rem`4
+
+type MatrixField r =
+    ( VectorSpace r
+    , Real r
+    , OrdField r
+    , Prim r
+    , Storable r
+    , Numeric.LinearAlgebra.Field r
```

Comment:
In the other matrix implementation, these constraints were needed because HMatrix required them.  It's probably possible to remove them here, which would be needed to remove the HMatrix dependency.